controllers/ibu_controller.go

```diff
@@ -109,53 +114,178 @@ func (r *ImageBasedUpgradeReconciler) Reconcile(ctx context.Context, req ctrl.Re
+					metav1.ConditionTrue,
+					"In progress",
+					ibu.Generation,
+				)
+				utils.SetStatusCondition(&ibu.Status.Conditions,
+					utils.ConditionTypes.Idle,
+					utils.ConditionReasons.InProgress,
+					metav1.ConditionFalse,
+					string(utils.GetInProgressConditionType(desiredStage)),
+					ibu.Generation)
+			}
+			nextReconcile, err = r.handleStage(ctx, ibu, desiredStage)
+			if err != nil {
+				return
+			}
+		}
 	}
+
 	// Update status
 	err = r.updateStatus(ctx, ibu)
 	return
 }
 
-func (r *ImageBasedUpgradeReconciler) handleCguFinalizer(
-	ctx context.Context, ibu *ranv1alpha1.ImageBasedUpgrade) (int, error) {
+func (r *ImageBasedUpgradeReconciler) handleStage(ctx context.Context, ibu *ranv1alpha1.ImageBasedUpgrade, stage ranv1alpha1.ImageBasedUpgradeStage) (nextReconcile ctrl.Result, err error) {
+	switch stage {
+	case ranv1alpha1.Stages.Idle:
+		nextReconcile, err = r.handleAbortOrFinalize(ctx, ibu)
+	case ranv1alpha1.Stages.Prep:
+		nextReconcile, err = r.handlePrep(ctx, ibu)
+	case ranv1alpha1.Stages.Upgrade:
+		nextReconcile, err = r.handleUpgrade(ctx, ibu)
+	case ranv1alpha1.Stages.Rollback:
+		nextReconcile, err = r.handleRollback(ctx, ibu)
+	}
+	return
+}
 
-	isCguMarkedToBeDeleted := ibu.GetDeletionTimestamp() != nil
-	if isCguMarkedToBeDeleted {
-		if controllerutil.ContainsFinalizer(ibu, utils.CleanupFinalizer) {
-			// Run finalization logic for cguFinalizer. If the finalization logic fails, don't remove the finalizer so
-			// that we can retry during the next reconciliation.
-			// err := utils.FinalMultiCloudObjectCleanup(ctx, r.Client, ibu)
-			// if err != nil {
-			// 	return utils.StopReconciling, err
-			// }
+func (r *ImageBasedUpgradeReconciler) handleAbortOrFinalize(ctx context.Context, ibu *ranv1alpha1.ImageBasedUpgrade) (nextReconcile ctrl.Result, err error) {
+	idleCondition := meta.FindStatusCondition(ibu.Status.Conditions, string(utils.ConditionTypes.Idle))
+	if idleCondition != nil && idleCondition.Status == metav1.ConditionFalse {
+		switch idleCondition.Reason {
+		case string(utils.ConditionReasons.Aborting):
+			nextReconcile, err = r.handleAbort(ctx, ibu)
+		case string(utils.ConditionReasons.AbortFailed):
+			nextReconcile, err = r.handleAbortFailure(ctx, ibu)
+		case string(utils.ConditionReasons.Finalizing):
+			nextReconcile, err = r.handleFinalize(ctx, ibu)
+		case string(utils.ConditionReasons.FinalizeFailed):
+			nextReconcile, err = r.handleFinalizeFailure(ctx, ibu)
+		}
+		if nextReconcile.Requeue == false {
+			utils.ResetStatusConditions(&ibu.Status.Conditions, ibu.Generation)
 		}
-		return utils.StopReconciling, nil
 	}
+	return
+}
 
-	// Add finalizer for this CR.
-	if !controllerutil.ContainsFinalizer(ibu, utils.CleanupFinalizer) {
-		controllerutil.AddFinalizer(ibu, utils.CleanupFinalizer)
-		err := r.Update(ctx, ibu)
-		if err != nil {
-			return utils.StopReconciling, err
+func isRollbackAllowed(ibu *ranv1alpha1.ImageBasedUpgrade) bool {
```

Comment:
Perhaps a description/comment can help people quickly understand when a rollback is allowed. Same as finalize and abort.

Reply:
Updated